### PR TITLE
ARROW-3648: [Plasma][Java] Add API to get metadata and data at the same time

### PR DESCRIPTION
**Pull request by guoyuhong**
Current Arrow Java Plasma client has no API to get the metadata and data together in one API call. If we split this process into two API calls, the object status could be different. Current observation shows that the first call could be empty(object not stored yet) while the second call will success but the metadata and data does not match.

### CODE REVIEW

**Comment by guoyuhong** [comment]
@wesm I have add the dependency to POM, but the CI failed with `Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/lang3/tuple/Pair`. Do you know how solve the problem?

**Comment by xhochy** [comment]
> @wesm I have add the dependency to POM, but the CI failed with `Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/lang3/tuple/Pair`. Do you know how solve the problem?

@guoyuhong Which dependency did you add to the POM? The current diff only shows some whitespace corrections.

**Comment by praveenbingo** [comment]
> @wesm I have add the dependency to POM, but the CI failed with `Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/lang3/tuple/Pair`. Do you know how solve the problem?

Hi,

It looks like you hand managing the class path in the [ci](https://github.com/apache/arrow/blob/master/ci/travis_script_plasma_java_client.sh#L36), i guess that is the reason the pair class was not showing up.

Can you make [this](https://github.com/apache/arrow/blob/master/java/plasma/src/test/java/org/apache/arrow/plasma/PlasmaClientTest.java) a java test instead (using @Test annotation) and let maven manage the test execution?

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java [hunk 1]
```diff
@@ -110,6 +109,33 @@ public void fetch(byte[][] objectIds) {
...
+
+    List<ObjectStoreData> ret = new ArrayList<>();
+    for (int i = 0; i < bufs.length; i++) {
+      ByteBuffer databuf = bufs[i][0];
+      ByteBuffer metabuf = bufs[i][1];
+      if (databuf == null) {
+        ret.add(new ObjectStoreData(null, null));
+      } else {
+        byte[] data = new byte[databuf.remaining()];
+        databuf.get(data);
```
Comment (praveenbingo) [inline java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: [just for my understanding] is there a reason we are making a copy of the data? do we want them to immutable by the caller?

**Comment by guoyuhong** [comment]
@xhochy I tried to add the `Pair` package:
```xml
         <dependency>
             <groupId>org.apache.commons</groupId>
             <artifactId>commons-lang3</artifactId>
             <version>3.4</version>
         </dependency>
```
I looks like there are some version conflict. So I discarded `Pair` and use the self-defined class `ObjectStoreData`.

File: java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java (hunk 1 above)
Comment (guoyuhong) [reply java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: In my understanding, the buffer is from Plasma shared memory. It is safe to make a copy and the user can release the shared buffer.  Add @pcmoritz to make sure whether my understanding is correct.

Comment (guoyuhong) [reply java/plasma/src/main/java/org/apache/arrow/plasma/PlasmaClient.java]: I notice that the other `get` function has the same copy operation.

**Comment by guoyuhong** [comment]
@praveenbingo I tried to make this a java unit test. But test failed with `java.lang.Exception: Please set plasma store path in env PLASMA_STORE`. If I set the Plasma Store path, there will 
be `PlasmaClientTest.start:50 » UnsatisfiedLink no plasma_java in java.library.path`. The dynamic lib is not installed yet. I think that is the reason why this test is not a java unit test. Can I keep this test as it is?

**Comment by praveenbingo** [comment]
> @praveenbingo I tried to make this a java unit test. But test failed with `java.lang.Exception: Please set plasma store path in env PLASMA_STORE`. If I set the Plasma Store path, there will
> be `PlasmaClientTest.start:50 » UnsatisfiedLink no plasma_java in java.library.path`. The dynamic lib is not installed yet. I think that is the reason why this test is not a java unit test. Can I keep this test as it is?

Sure either ways is ok, but if you want to pass the library path that can be passed as an arg to mvn too as described [here.](https://stackoverflow.com/questions/35366035/set-java-library-path-for-testing)

**Comment by guoyuhong** [comment]
@praveenbingo Thanks for the information!

**Comment by guoyuhong** [comment]
@xhochy Do you have any comment?

**Comment by guoyuhong** [comment]
@robertnishihara Hi, Robert, could you take a look at this PR?

**Review by robertnishihara — APPROVED** [review]
This looks good to me, thanks @guoyuhong!

**Pull request closed without merging**